hoomd/md/AlchemyData.h

```diff
@@ -0,0 +1,191 @@
+    protected:
+    bool m_attached;
+    std::shared_ptr<const ExecutionConfiguration>
+        m_exec_conf;                 //!< Stored shared ptr to the execution configuration
+    std::shared_ptr<Compute> m_base; //!< the associated Alchemical Compute
+    };
+
+struct AlchemicalMDParticle : AlchemicalParticle
+    {
+    AlchemicalMDParticle(std::shared_ptr<const ExecutionConfiguration> exec_conf)
+
+        : AlchemicalParticle(exec_conf) {};
+
+    void inline zeroForces()
+        {
+        ArrayHandle<Scalar> h_forces(m_alchemical_derivatives,
+                                     access_location::host,
+                                     access_mode::overwrite);
+        memset((void*)h_forces.data, 0, sizeof(Scalar) * m_alchemical_derivatives.getNumElements());
+        }
+
+    void resizeForces(unsigned int N)
+        {
+        GlobalArray<Scalar> new_forces(N, m_exec_conf);
+        m_alchemical_derivatives.swap(new_forces);
+        }
+
+    void zeroNetForce(uint64_t timestep)
+        {
+        zeroForces();
+        m_timestep_net_force.first = timestep;
+        }
+
+    void setNetForce()
+        {
+        Scalar netForce(0.0);
+        ArrayHandle<Scalar> h_forces(m_alchemical_derivatives,
+                                     access_location::host,
+                                     access_mode::read);
+        for (unsigned int i = 0; i < m_alchemical_derivatives.getNumElements(); i++)
+            netForce += h_forces.data[i];
+        netForce /= Scalar(m_alchemical_derivatives.getNumElements());
+        m_timestep_net_force.second = netForce;
+        }
+
+    Scalar getNetForce(uint64_t timestep)
+        {
+        assert(m_timestep_net_force.first == timestep);
+        return m_timestep_net_force.second;
+        }
+
+    Scalar getNetForce()
+        {
+        return m_timestep_net_force.second;
+        }
+
+    void setMass(Scalar new_mass)
+        {
+        mass.x = new_mass;
+        mass.y = Scalar(1.) / new_mass;
+        }
+
+    Scalar getMass()
+        {
+        return mass.x;
+        }
+
+    Scalar getValue()
+        {
+        return value;
+        }
+
+    pybind11::array_t<Scalar> getDAlphas()
+        {
+        ArrayHandle<Scalar> h_forces(m_alchemical_derivatives,
+                                     access_location::host,
+                                     access_mode::read);
+        return pybind11::array(m_alchemical_derivatives.getNumElements(), h_forces.data);
+        }
+
+    Scalar momentum = 0.; // the momentum of the particle
+    Scalar2 mass
+        = make_scalar2(1.0,
+                       1.0); // mass (x) and it's inverse (y) (don't have to recompute constantly)
+    Scalar mu = 0.;          //!< the alchemical potential of the particle
+    GlobalArray<Scalar> m_alchemical_derivatives; //!< Per particle alchemical forces
+    protected:
+    // the timestep the net force was computed and the netforce
+    std::pair<uint64_t, Scalar> m_timestep_net_force;
```

Comment:
The original code author designed the alchemical integration methods to operate with periods greater than 1. The `uint64_t` tracks the timestep the force was computed for use with an assert to ensure that the alchemical derivatives were computed on the correct timestep.

I don't understand it either as I think that `PotentialPairAlchemical` computes the alchemical forces on all timsteps...

Reply:
It has been a while... but Josh is correct about the time keeping portion and asserts because of the different timestep scales. It's pretty important to get right with the Trotter factorization because obviously if you get to your next start 1/2 timestep and didn't just wrap up your end 1/2 timestep... you've got a problem. It could perhaps be rewritten into a test rather than this tag-along variable. However I intended it to be possible to add some warnings or checks to make sure that an alchemical particle isn't disowned by an integrator before it's completed a full alchemical step.

Reply:
@joaander as to your second point. `PotentialPairAlchemical` shouldn't compute much for each timestep if the masking aspect is working properly and the Trotter factorization power is greater than 2.

Reply:
> @joaander as to your second point. `PotentialPairAlchemical` shouldn't compute much for each timestep if the masking aspect is working properly and the Trotter factorization power is greater than 2.

OK, I see it now.

Reply:
Resolving this. If you feel strongly that it should be written a different way, feel free to fix it yourself.